pkg/sql/sqlbase/structured.go

```diff
@@ -2304,6 +2314,19 @@ func (desc *TableDescriptor) validateTableIndexes(columnNames map[string]ColumnI
 					index.Name, index.Sharded.Name)
 			}
 		}
+
+		if index.IsPartial() && SchemaExprPartialIndexPredicateValidateHook != nil {
+			expr, err := parser.ParseExpr(index.Predicate)
+			if err != nil {
+				return err
+			}
+
+			immutableDesc := NewImmutableTableDescriptor(*desc)
+			_, err = SchemaExprPartialIndexPredicateValidateHook(ctx, immutableDesc, expr, semaCtx)
```

Comment:
I need to plumb a `*tree.SemaContext` here in order to support user-defined types in partial index predicates—the TypeResolver is required to validate the predicate expression. I started the plumbing in the 2nd commit (WIP), but was running into some places where a `semaCtx` doesn't seem to exist, like `catalogkv.go`.

@rohany does all this plumbing of `semaCtx` seem _right_ to you? Is there a better way?

Reply:
I don't think you need to do this. In ValidateCrossReferences I am already using `GetAllReferencedTypeIDs` which includes types in partial index predicates and verifying that all of those types exist.

Reply:
I see. My goal wasn't specifically to check that the reference types within partial index predicates exist, but more broadly to validate partial index predicate expressions are valid. So I thought it would be safest to use the same validator that validates these predicates when the indexes are created. That validator uses `sqlbase.SanitizeVarFreeExpr`, which performs type-checking, and it fails if we don't have a TypeResolver on the semaCtx that can resolve UDTs.

So maybe the question is do we want to perform the same validation here that we do on index creation?

If the answer is yes, I'll need some sort of TypeResolver, whether by plumbing a semaCtx into this function, or creating a new implementation of the interface that is attached to a new semaCtx.

If the answer is no, then what level of validation should we be performing here?

Reply:
I see. This might be tricky, I'm not sure that we pass down valid SemaCtx's all the way to the methods that use validate here. An option is to implement a new TypeResolver backed by a txn that just looks up the types by ID -- i don't think that there should be any names to resolve in these expressions, as that would be a bug. 

On the meta level question, I'm not sure what is the goal level of validation. I think that "as much as possible" is the answer, so we should try our best here.

Reply:
If this plumbing is particularly painful, I think it's acceptable to only validate that the referenced types exist. I would say that referential integrity is the main thing we're worried about here, unless either of you can think of scenarios where the UDT would exist but the expression had somehow ceased to be valid.